packages/react-devtools-shared/src/devtools/views/hooks.js

```diff
@@ -300,3 +302,29 @@ export function useSubscription<Value>({
 
   return state.value;
 }
+
+// provides highlighting and clearing highlights of elements based on ID
+export function useNativeElementHighlighter() {
+  const store = useContext(StoreContext);
+  const bridge = useContext(BridgeContext);
+
+  const highlightNativeElement = (id: number) => {
```

Comment:
The previous version of `highlightNativeElement` was memoized with `useCallback`. This one isn't, which would cause the effect it's passed to in the `Tree` component to be run every render.

Reply:
In this case, it probably is good to wrap it in a `useCallback` as before. 

But I saw several places where `useCallback` is wrapping simple functions like 
```js
const handleBlur = useCallback(() => setTreeFocused(false), []);
``` 
From [When to useMemo and useCallback](https://kentcdodds.com/blog/usememo-and-usecallback) article I understand that in these cases it doesn't give better performance. 
I want to know your opinion on this topic, to understand if I should use `useCallback` in such cases.

Reply:
Interesting,
I was wondering if it is possible to use profiler devTools while debugging profiler devTools? :D

I did a small study:
* used both codes (with `useCallback` and without `useCallback`) on a real project
* Then tried to highlight different elements by moving the mouse randomly
* Put `console.time/console.timeEnd` before calling `highlightNativeElement` and after sending to bridge.
* Gathered 800 data points and averaged them

And both cases almost took a similar amount of time on my Mac:
1. **with `useCallback`:** 0.01529866435s** (Average of 800 highlights)
1. **without `useCallback`:** 0.01466493762s** (Average of 800 highlights)

The results basically even show that using `useCallback` has a minor negative (or even negligible to be honest) effect on performance. I guess the trade-off of using `useCallback` is almost equivalent to the calculation done in the function. 

Either way, there is a flaw in my case study, or as a usual practice we prefer to use `useCallback` on all/most methods in the project, please let me know, I will add it to the method:))

Reply:
> I was wondering if it is possible to use profiler devTools while debugging profiler devTools? 

Yes. You can use the DevTools extension to debug/profile the test shell (which runs an inline version of the DevTools).

My concern was that it would break memoization and cause more/unnecessary rendering. I'm not sure I understand how the console timing you describe above measures this. I think you would need to use a profiler to measure it.

Let me take another look sometime later this morning.

Reply:
> My concern was that it would break memoization and cause more/unnecessary rendering.

I think that even if with `useCallback` it takes more time to evaluate between `console.time/console.timeEnd` the purpose is to no rerender. But still, this means that we should use `useCallback` with every function that we pass to children (which sounds strange). 
So probably the best way is to first pass function without `useCallback` and if there are performance issues then optimize. 

As I see in [Tree.js](https://github.com/facebook/react/blob/master/packages/react-devtools-shared/src/devtools/views/Components/Tree.js) every child prop function is wrapped with `useCallback`. 
So interesting question is: What is the performance without `useCallback`? Is it good enough to not `useCallback`?

Reply:
There are some unnecessary `useCallback` wrappers in this repo. DevTools was my the first app I wrote that used hooks, so there are some mistakes :smile:

In this particular case, `highlightNativeElement` *used to be a stable function*. Because it gets passed through to `itemData`, it seemed preferable for it to remained stable-ish to avoid unnecessarily re-rendering a lot of chart nodes which are memoized: https://github.com/facebook/react/blob/90be006da8e231279151e7b1518bb64c62e26851/packages/react-devtools-shared/src/devtools/views/Profiler/CommitFlamegraphListItem.js#L118

Because `store` and `bridge` are stable, callbacks that only use those values should be stable too.

I'm not sure how important this turns out to be in practice. This is why I had suggested React profiling :smile:

Reply:
Interesting, that makes sense:)
About this case that we were discussing and also other use cases of using the `useCallback` and memoizations, I can try to go through the code, and do some additional checks to see how these changes affect the number of updates on the page and also probably the performance.  Would it be something that we are interested in?

Reply:
About this one, I checked it out, there is no additional re-renders fortunately, as you mentioned store and bridge are stable. So I think it makes sense to keep highlighter without `useCallback`